### PR TITLE
feat(List): add column chooser for VList and switch toolbar buttons aspect

### PR DESCRIPTION
**Pull request by Nigui**
**What is the problem this PR is trying to solve?**

Before now, when you wanted to add a column chooser button without toolbar, you had to handle column visibility by hand. That period is over. VList component will handle it for you. 

It adds optional column chooser button in table header of a VList displayed in table mode. 

column-chooser-vlist-table

It also reworks toolbar display mode toggle and column chooser button to use `ButtonIcon...` to keep consistency with column chooser in header. 

toolbar-buttons

**What is the chosen solution to this problem?**

Component user can add column chooser feature to its `VList` just passing `columnChooser` boolean property

```
<List.Manager
	id="my-list"
	collection={collection}
>
	<List.VList id="my-vlist" columnChooser>
		...
	</List.VList>
</List.Manager>
```

User can also pass props to `columnChooser` like original `columnChooser` property in `List` (e.g. `nbLockedLeftItems`, `columns`, `onSubmit`,...).

Change is fully backward-compatible.

**Please check if the PR fulfills these requirements**

- [x] The PR have used `yarn changeset` to a request a release from the CI if wanted.
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

**Comment by ncomont** [comment]
(changeset is missing)

File: packages/components/src/List/ListComposition/ColumnChooser/ColumnChooserHeaderButton.component.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import PropTypes from 'prop-types';
+import { ButtonIconFloating } from '@talend/design-system';
+
+import ColumnChooser from './ColumnChooser.component';
+import ColumnChooserButton from '../../Toolbar/ColumnChooserButton';
+import theme from './ColumnChooserHeaderButton.scss';
+
+export default function ColumnChooserHeaderButton(props) {
+	return (
+		<div className={theme.columnChooserHeaderButtonContainer}>
```
Comment (ncomont) [inline packages/components/src/List/ListComposition/ColumnChooser/ColumnChooserHeaderButton.component.js:11]: we usually use kebab case for class names

```suggestion
		<div className={theme['column-chooser-header-button-container']}>
```

File: packages/components/src/List/ListComposition/ColumnChooser/ColumnChooserHeaderButton.component.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+			<ColumnChooser {...props} buttonRenderer={ButtonIconFloating} />
+		</div>
+	);
+}
+
+ColumnChooserHeaderButton.propTypes = PropTypes.oneOfType(
```
Comment (ncomont) [inline packages/components/src/List/ListComposition/ColumnChooser/ColumnChooserHeaderButton.component.js:17]: not sure of these proptypes 🤔

File: packages/components/src/List/ListComposition/DisplayMode/ListDisplayMode.component.js [hunk 3]
```diff
@@ -68,18 +19,13 @@ function ListDisplayMode({ children, displayModesOptions, id, onChange, selected
...
-					key={displayModeOption}
-					onSelect={onSelect}
-				/>
-			))}
-		</div>
+		<DisplayModeToggle
+			id={id}
+			mode={selectedDisplayMode || displayMode}
+			displayModes={displayModesOptions}
+			onChange={(event, value) => onSelect(event, value)}
```
Comment (ncomont) [inline packages/components/src/List/ListComposition/DisplayMode/ListDisplayMode.component.js:26]:
```suggestion
			onChange={onSelect}
```

File: packages/components/src/List/ListComposition/utils.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export const columnsFromChildrens = children => {
+	return Array.isArray(children)
+		? children.filter(column => column.props?.dataKey).map(column => column.props)
```
Comment (ncomont) [inline packages/components/src/List/ListComposition/utils.js:3]: could the data key be '0' ?

```suggestion
		? children.filter(column => column.props?.dataKey != null).map(column => column.props)
```

File: packages/components/src/List/Toolbar/Toolbar.scss [hunk 5]
```diff
@@ -90,7 +92,6 @@ $tc-list-toolbar-height: 7rem !default;
 	}
 
 	:global(li.separated button[data-feature='column-chooser.open']) {
-		margin: 0;
-		padding-right: 0;
+		margin-left: tokens.$coral-size-m;
```
[the remark this answers is not in the document]
Comment (Nigui) [reply packages/components/src/List/Toolbar/Toolbar.scss:95]: Sorry they existed before in design-tokens 2.0.0
https://github.com/Talend/ui/blame/22fee906dc8756be529b588de9c20d69d24f94d5/packages/design-tokens/src/_tokens.scss#L226
Now they're named `$coral-sizing-*`. I'll update using `-spacing-` which is more appropriate for `margin-left` prop. 
https://github.com/Talend/ui/blame/master/packages/design-tokens/src/_tokens.scss#L223-L237

File: packages/components/src/List/ListComposition/ColumnChooser/ColumnChooserHeaderButton.component.js (hunk 2 above)
Comment (Nigui) [reply packages/components/src/List/ListComposition/ColumnChooser/ColumnChooserHeaderButton.component.js:17]: What do you mean? Not a good practice to have property with 2 possible types? Or you're not sure about current types?

File: packages/components/src/List/ListComposition/DisplayMode/ListDisplayMode.component.js (hunk 3 above)
Comment (Nigui) [reply packages/components/src/List/ListComposition/DisplayMode/ListDisplayMode.component.js:26]: I had the `jsx-no-bind` linter error https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md
But I transformed function declaration to make it pass :)

Comment (Nigui) [reply packages/components/src/List/ListComposition/DisplayMode/ListDisplayMode.component.js:26]: I had the `jsx-no-bind` linter error https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md
But I transformed function declaration to make it pass :)

File: packages/components/src/List/ListComposition/utils.js (hunk 4 above)
Comment (Nigui) [reply packages/components/src/List/ListComposition/utils.js:3]: This code comes from https://github.com/Talend/ui/pull/3842/files#diff-8d68d34558e914cec5b15cc6aac0ffddb5fdacc6885de7eec205647f55f38cc3L26
but we can improve it for sure ! Thank you

File: packages/components/src/List/Toolbar/Toolbar.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (frassinier) [reply packages/components/src/List/Toolbar/Toolbar.scss]: This is pretty new indeed :)

**Review by ncomont — APPROVED** [review]

**Comment by ncomont** [comment]
(minor) pop over does not follow the button when resizing the window

image

**Comment by ncomont** [comment]
The button is above the sort order indicator 

image

**Comment by ncomont** [comment]
how should I focus this button using my keyboard ? from the header of the last column, when I hit `tab`, the first cell of the first column is focused. Is it accessible ?

**Review by ncomont — CHANGES_REQUESTED** [review]
see comments

**Comment by ncomont** [comment]
valid and validWithIcon columns are empty in sb 
image

**Comment by ncomont** [comment]
legacy column chooser button is blue by default : 
image
not on master :
image

**Comment by ncomont** [comment]
if my memories are good, the button should be vertically centered. here, the shadow is taken into account :
image

**Comment by Nigui** [comment]
> if my memories are good, the button should be vertically centered. here, the shadow is taken into account : image

Unfortunately there is no design-token with `0.6rem`, but I could set without token

----

> valid and validWithIcon columns are empty in sb image

fixed 👍 

----

> (minor) pop over does not follow the button when resizing the window
> 
> image

It's not related to current modification. We could create a new ticket to handle this bad UX. But maybe switching to design-system's popover would fix because it uses [reakit](https://reakit.io/docs/popover/) (and not react-bootstrap).

----- 

> how should I focus this button using my keyboard ? from the header of the last column, when I hit `tab`, the first cell of the first column is focused. Is it accessible ?

Yes it's the first element when you enter the table. Right before the first header cell.

-----

For the 2 other comments on column chooser button color and spacing in toolbar and the floating button in header hiding sort icon, I'll ask confirmation from UX team

**Review by ncomont — APPROVED** [review]

**Review by romainseb — APPROVED** [review]

**Comment by jsomsanith-tlnd** [comment]
I have the problem with this design. The button is floating but still located in the last header, making it feel that it's an action on the last column.

Furthermore, we can't see the sort caret anymore in some cases:
image
image


Can you fix the caret issue and double check with your UX for the placement ?

**Pull request merged**